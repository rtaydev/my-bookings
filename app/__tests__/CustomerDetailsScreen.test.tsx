import React from "react";
import { render } from "@testing-library/react-native";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import CustomerDetailsScreen from "../customer-details";
import bookingsReducer from "@/store/slices/bookingSlice"; // Adjust this import path as needed

// Mock the expo-router hook
jest.mock("expo-router", () => ({
  useLocalSearchParams: jest.fn(() => ({
    surname: "Test",
    bookingReference: "123",
  })),
}));

const createTestStore = (initialState) => {
  return configureStore({
    reducer: {
      bookings: bookingsReducer,
    },
    preloadedState: {
      bookings: initialState,
    },
  });
};

describe("CustomerDetailsScreen", () => {
  beforeEach(() => {
    // Mock the fetchBookings action
    jest
      .spyOn(require("@/store/slices/bookingSlice"), "fetchBookings")
      .mockImplementation(() => ({ type: "fetchBookings" }));
  });

  it("renders loading indicator when loading", () => {
    const store = createTestStore({
      loading: true,
      error: null,
      futureBookings: [],
      historicBookings: [],
    });
    const { getByTestId } = render(
      <Provider store={store}>
        <CustomerDetailsScreen />
      </Provider>,
    );
    expect(getByTestId("loading-indicator")).toBeTruthy();
  });

  it("renders error message when there is an error", () => {
    const store = createTestStore({
      loading: false,
      error: "Error fetching bookings",
      futureBookings: [],
      historicBookings: [],
    });
    const { getByTestId } = render(
      <Provider store={store}>
        <CustomerDetailsScreen />
      </Provider>,
    );
    expect(getByTestId("error-message")).toBeTruthy();
    expect(getByTestId("error-message").props.children).toEqual([
      "Error: ",
      "Error fetching bookings",
    ]);
  });

  it("renders future and historic bookings", () => {
    const store = createTestStore({
      loading: false,
      error: null,
      futureBookings: [{ id: "1", title: "Future Booking" }],
      historicBookings: [{ id: "2", title: "Historic Booking" }],
    });
    const { getByTestId } = render(
      <Provider store={store}>
        <CustomerDetailsScreen />
      </Provider>,
    );
    expect(getByTestId("future-booking-1")).toBeTruthy();
    expect(getByTestId("historic-booking-2")).toBeTruthy();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
