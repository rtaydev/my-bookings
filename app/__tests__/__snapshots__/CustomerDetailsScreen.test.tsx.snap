// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerDetailsScreen matches the snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
      "padding": 20,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 24,
        "fontWeight": "bold",
        "marginBottom": 20,
      }
    }
  >
    Your Ferry Reservations
  </Text>
  <RCTScrollView
    data={
      [
        {
          "id": "1",
          "title": "Future Booking",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#f9f9f9",
              "borderRadius": 5,
              "marginVertical": 8,
              "opacity": 1,
              "padding": 20,
            }
          }
          testID="future-booking-1"
        >
          <Text
            style={
              {
                "fontSize": 18,
                "fontWeight": "bold",
              }
            }
          >
            Future Booking
          </Text>
          <Text
            style={
              {
                "color": "#666",
                "fontSize": 14,
              }
            }
          >
             at 
          </Text>
          <Text
            style={
              {
                "color": "#666",
                "fontSize": 14,
              }
            }
          >
             passengers, 
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <Text
    style={
      {
        "fontSize": 24,
        "fontWeight": "bold",
        "marginBottom": 20,
      }
    }
  >
    Your Historic Reservations
  </Text>
  <RCTScrollView
    data={
      [
        {
          "id": "2",
          "title": "Historic Booking",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#f9f9f9",
              "borderRadius": 5,
              "marginVertical": 8,
              "opacity": 1,
              "padding": 20,
            }
          }
          testID="historic-booking-2"
        >
          <Text
            style={
              {
                "fontSize": 18,
                "fontWeight": "bold",
              }
            }
          >
            Historic Booking
          </Text>
          <Text
            style={
              {
                "color": "#666",
                "fontSize": 14,
              }
            }
          >
             at 
          </Text>
          <Text
            style={
              {
                "color": "#666",
                "fontSize": 14,
              }
            }
          >
             passengers, 
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
